pxd_fastpath.c

```diff
@@ -515,6 +515,49 @@ void pxd_fastpath_adjust_limits(struct pxd_device *pxd_dev, struct request_queue
 	disableFastPath(pxd_dev, false);
 }
 
+// reset device called during device cleanup actions from any internal state.
+// consider node wipe, device remove while suspended etc.
+void pxd_fastpath_reset_device(struct pxd_device *pxd_dev)
+{
+	struct pxd_fastpath_extension *fp = &pxd_dev->fp;
+	struct pxd_context *ctx = pxd_dev->ctx;
+	struct fuse_conn *fc = &ctx->fc;
+	struct fuse_req *req;
+
+	if (!fastpath_enabled(pxd_dev)) {
+		return;
+	}
+
+	disableFastPath(pxd_dev, true);
+	// resume from userspace IO suspends after px restarts
+	pxd_request_resume(pxd_dev);
```

Comment:
Would px-storage resume this IO in which case there might be some circular dependency again.
This path may get invoked when px-storage is down. Reset_device should just kill all requests for the device and unblock.

Reply:
There are only 3 cases where this method to cleanup and abort all IOs in fastpath gets called.

1/ pxd device remove
2/ px abort context when px userspace times out connecting to driver.
3/ node wipe gets run.

all enqueued IOs gets aborted with EIO when this gets run. There is no circular dependency from here.

Reply:
Device remove is called from px-storage. If px-fuse call back into px-storage and storage is down (startup force detach case) it would deadlock.

Reply:
in this code path, px-fuse will not callback into px-storage, all pending IOs on that volume that was queued while volume is suspended, would be failed with EIO.